docs/testing-guidelines.md

```diff
@@ -0,0 +1,122 @@
+> The minimal and least fragile way to test all features of an application that could be broken from how we deploy it.
+
+#### <a name="key-principles"></a>**Key Principles**
+- Focus on testing the deployment, not the application itself (that's the upstream developer's job).
+- Ensure that our deployment method doesn't break key functionalities.
+- Test integration with UDS core components (e.g., Istio, Keycloak).
+
+#### <a name="implementation-guidelines"></a>**Implementation Guidelines**
+- Design tests that cover essential functionality without being overly complex.
+- Focus on deployment-related issues (e.g., network policies, SSO access, cluster resource access).
+- Handle limitations due to licensing or other constraints by documenting them and implementing what testing is possible.
+
+#### <a name="tools"></a>**Tools**
+- UI Testing: Playwright, Cypress
+- Non-UI Testing: Jest
+- Custom Scripts: Bash or other scripting languages as needed
+
+## <a name="upgrade-testing"></a>Upgrade Testing
+Ensure that the current development package works when deployed over the previously released version.
+
+#### <a name="key-considerations"></a>**Key Considerations**
+- Test data migration and persistence.
+- Verify that configurations are properly carried over or updated.
+- Check for any breaking changes in APIs or integrations.
+
+## <a name="linting-and-static-analysis"></a>Linting and Static Analysis
+
+Use appropriate tooling to ensure code quality and consistency.
+
+#### <a name="recommended-tools"></a>**Recommended Tools**
+- [YAML Linting](https://github.com/adrienverge/yamllint): `yamllint`
+- [Zarf Package Linting](https://docs.zarf.dev/commands/zarf_dev_lint/): `zarf dev lint`
+- [Shellcheck](https://www.shellcheck.net/) for any bash scripts.
+
+## <a name="best-practices"></a>**Best Practices**
+
+1. **Consistency**: Use similar testing approaches across different UDS packages where possible.
+2. **Documentation**: Clearly document what each test is checking and why.
+3. **Maintenance**: Keep tests updated as the application and deployment methods evolve.
+4. **CI/CD Integration**: Incorporate all types of tests into your CI/CD pipeline.
+5. **Error Handling**: Ensure tests fail gracefully and provide clear error messages.
+6. **Idempotency**: Ensure tests produce consistent results regardless of execution order. Design them to handle dynamic and asynchronous workloads without compromising output integrity.
+7. **Performance Consideration**: Balance thoroughness with rapid developer feedback:
+    - **Single package tests:** Aim for <15 minutes runtime
+    - **Full CI jobs** (e.g., deploy and test): Maximum 30 minutes
+    - **Large integration tests** (multiple packages): Longer durations acceptable
+
+    > **Note:** Strive for individual package tests to complete in under 10 minutes when possible.
+
+
+## <a name="best-practice-examples"></a>**Best Practice Examples**
```

Comment:
as a structural / stylistic question - should we pull the examples up as sub-sections to the sections where we define them?

Reply:
I think I follow. Rather than a "Definition" and then a separate "Best Practice Example" for each definition, basically melding the two? I agree, it would be more cohesive to make these sections blended.

Reply:
https://github.com/defenseunicorns/uds-common/pull/205/commits/60103b9bfefa1066caeffec5542fee9d3aa1c365 this aimed to combine the 2 sections. I reordered it a bit cause I think it flowed better that way, let me know if you'd like me to change any piece of it.

I do think it is kind of crowded? But ultimately I think it's a good change.